sys/sys/linker.h

```diff
@@ -219,6 +219,14 @@ void linker_kldload_unbusy(int flags);
 
 #endif	/* _KERNEL */
 
+/*
+ * ELF file types
+ */
+#define KERNTYPE_MB	"elf multiboot kernel"
```

Comment:
I still want these to be constant objects instead of literals hidden under a macro.

Reply:
No, of course it cannot be shared between kernel and loader.  But inside kernel, and inside loader, you should use a symbol pointing to the same string.  This would avoid spending memory in both kernel and loader for copies of the same string literal.

For kernel, it is nice not to waste memory.  For loader, we have issues with the BIOS limitations, so even a byte saved matters.

Reply:
afaik this shouldn't be a problem; we compile with constant merging enabled, so all the string literals should point to the same place.
For GCC, -fmerge-constants is implicitly enabled at -O1 or higher.
For Clang, as far as I can tell, there isn't a way to disable it.
The merging itself is of course done by the linker.

```
$ uname -r
14.1-RELEASE
$ strings /boot/loader | grep 'elf kernel'
elf kernel
$ strings /boot/loader.efi | grep 'elf kernel'
elf kernel
$ strings /boot/kernel/kernel | grep 'elf kernel'
elf kernel
```

I do see your point though.

The reason I went for macros in the first place was because I couldn't find a good place for the global in the loader. It can't be in load_elf.c since that gets compiled twice and would cause a duplicate symbol error. We could wrap it in __elfN, but then we'd have two symbols for no reason and we would need to use __elfN in other files that reference the string. Outside of load_elf.c and load_elf_obj.c, I'm not sure where to put it.

Reply:
I am curious, and do not have time to dig it myself.  How do linkers do the deduplication of the string literals?  I do not see any way for this to work except putting each literal into a dedicated section with specific attributes.

That said, I do not think that finding a place is a big problem.  Just pointing a finger into the wall, I suggest stand/common/boot.c.  If you do not like it, you can add a new file into libsa or common, to collect literals.

Reply:
> I am curious, and do not have time to dig it myself. How do linkers do the deduplication of the string literals? I do not see any way for this to work except putting each literal into a dedicated section with specific attributes.

Almost, but multiple string literals can be in the same section. e.g a file consisting of only
```c
const char *foo = "bar";
const char *bar = "foo";
```
puts both "bar" and "foo" in `.rodata.str1.1`. Using readelf, these are the section's flags: `ALLOC, MERGE, STRINGS`
Why the number at the end? Apparently its so that it can create another section is in case we need different alignment for another set of strings, but I didn't look too much into that.

The `STRINGS` flag means that the section consists of only null-terminated strings, and the `MERGE` flag instructs the linker to attempt to merge them. I'm not sure how exactly the actual merging happens though, that's about where the depth of my knowledge on this ends.

> That said, I do not think that finding a place is a big problem. Just pointing a finger into the wall, I suggest stand/common/boot.c. If you do not like it, you can add a new file into libsa or common, to collect literals.

Hm, I'll have another look around the loader, I feel like I dismissed the idea of a global too quickly. common/boot.c isn't a bad suggestion :)